src/org/testKitGen/TestDivider.java

```diff
@@ -60,41 +73,234 @@ public static void addTests(File playlistXML) {
+					dataFromTRSS = true;
+				} else {
+					actualTestMap.put(test, 0);
+					testsNotFound.add(test);
+				}
+			}
+			for (Map.Entry<String, Integer> entry : actualTestMap.entrySet()) {
+				testTimeQueue.offer(entry);
+			}
+		}
+		System.out.println("\nTEST DURATION");
+		System.out.println("====================================================================================");
+		int totalNum = regularTests.size() + effortlessTests.size();
+		System.out.println("Total number of tests searched: " + totalNum);
+		if (dataFromTRSS) {
+			int foundNum = testTimeQueue.size() - testsNotFound.size();
+			System.out.println("Number of test durations found: " + foundNum);
+			System.out.println("Top slowest tests: ");
+			List<Map.Entry<String, Integer>> lastTests = new ArrayList<>();
+			for (int i = 0; i < Math.min(foundNum, 3); i++) {
+				lastTests.add(testTimeQueue.poll());
+			}
+			for (int i = 0; i < lastTests.size(); i++) {
+				System.out.println("\t" + formatTime(lastTests.get(i).getValue()) + " " + lastTests.get(i).getKey());
+				testTimeQueue.offer(lastTests.get(i));
+			}
+			if (!testsNotFound.isEmpty()) {
+				System.out.println("Following test duration data cannot be found: ");
+				System.out.println("(Default duration assigned, running tests: " + defaultAvgTestTime / 1000 + "s; skipped/disabled tests: 0s.)");
+				System.out.println(testsNotFound.toString());
+			}
+		} else {
+			System.out.println("No test duration data found.");
+			System.out.println("(Default duration assigned, running tests: " + defaultAvgTestTime / 1000 + "s; skipped/disabled tests: 0s.)");
+
+		}
+		System.out.println("====================================================================================");
+	}
+
+	private static String formatTime(int milliSec) {
+		int sec = milliSec / 1000;
+		String duration = String.format("%02dm%02ds", sec / 60, sec % 60);
+		return duration;
+	}
 
 	public static void generateLists() {
-		if (Options.getNumOfMachine() == null) {
+		if (regularTests.size() == 0 && effortlessTests.size() == 0) {
+			System.out.println("No tests found for target: " + TestTarget.getTestTarget());
+			System.out.println("No parallel test lists generated.");
+			return;
+		}
+
+		if (TestTarget.isDisabled()) {
+			// TRSS does not contain test duration for running disabled test at this moment
+			System.out.println("Warning: Test duration data cannot be found for executing disabled target.");
+			System.out.println("(Default duration assigned, running tests: " + defaultAvgTestTime / 1000 + "s; skipped tests: 0s.)");
+		} else {
+			buildTestTimeQueue();
+		}
+	
+		if (Options.getnumOfMachines() == null) {
 			if (Options.getTestTime() == null) {
 				divideOnMachineNum(1);
 			} else {
 				divideOnTestTime(Options.getTestTime());
 			}
 		} else {
-			divideOnMachineNum(Options.getNumOfMachine());
+			divideOnMachineNum(Math.min(Options.getnumOfMachines(), regularTests.size() + effortlessTests.size()));
```

Comment:
What is the `effortlessTests `? skipped or disabled tests? If so, should we do `Math.min(Options.getnumOfMachines(), regularTests.size())` here?

Reply:
Yes, they are skip/disabled tests. We can only consider the regular tests here. It will then be `Math.min(1, Math.min(Options.getnumOfMachines(), regularTests.size()))` Two options here, it is either to have the system smart or have the system do what user ask it to do. For example, we only have 3 effortlessTests, the user give NumOfMachines=3. With effortlessTests it will be break into 3 jobs. Without effortlessTests, it will be 1 job but still finish quick.

Reply:
updated to `Math.max(1, Math.min(Options.getnumOfMachines(), regularTests.size()))`